File: backend/app/modules/fund_nav/schemas/market.py
from __future__ import annotations

from datetime import date, datetime
from decimal import Decimal
from typing import Literal

from pydantic import BaseModel


class MarketQuoteOut(BaseModel):
    asset_code: str
    asset_name: str | None = None
    asset_type: str
    market: str | None = None
    trade_date: date
    quote_time: datetime
    latest_price: Decimal | None = None
    prev_close: Decimal | None = None
    change_rate: Decimal | None = None
    source: str

    model_config = {"from_attributes": True}


class IndexQuoteSourceStatusOut(BaseModel):
    id: int
    source_key: str
    source_name: str
    source_description: str | None = None
    source_type: str
    source_type_label: str
    exclude_rule_type: str
    exclude_rule_value: str | None = None
    priority: int
    enabled: int
    success_count: int
    failure_count: int
    consecutive_failures: int
    success_rate: Decimal | None = None
    failure_rate: Decimal | None = None
    effective_priority: Decimal
    auto_disabled_until: datetime | None = None
    last_success_at: datetime | None = None
    last_failure_at: datetime | None = None
    last_error: str | None = None
    status_label: str


class IndexQuoteSourceRuleIn(BaseModel):
    enabled: Literal[0, 1] | None = None
    source_description: str | None = None
    exclude_rule_type: str = "none"
    exclude_rule_value: str | None = None


class IndexQuoteSymbolOut(BaseModel):
    id: int
    index_code: str
    source_key: str
    quote_symbol: str | None = None
    supported: int
    description: str | None = None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class IndexQuoteSymbolIn(BaseModel):
    index_code: str
    source_key: str
    quote_symbol: str | None = None
    supported: int = 1
    description: str | None = None


class IndexQuoteSymbolPageOut(BaseModel):
    items: list[IndexQuoteSymbolOut]
    total: int
    limit: int
    offset: int

File: frontend/src/modules/fund_nav/api/market.ts
import { apiClient } from '../../../api/client'

export interface RefreshMarketResult {
  refreshed: boolean
  quote_count: number
}

export interface IndexQuoteSourceStatus {
  id: number
  source_key: string
  source_name: string
  source_description?: string | null
  source_type: string
  source_type_label: string
  exclude_rule_type: string
  exclude_rule_value: string | null
  priority: number
  enabled: number
  success_count: number
  failure_count: number
  consecutive_failures: number
  success_rate: string | null
  failure_rate: string | null
  effective_priority: string
  auto_disabled_until: string | null
  last_success_at: string | null
  last_failure_at: string | null
  last_error: string | null
  status_label: string
}

export interface IndexQuoteSymbol {
  id: number
  index_code: string
  source_key: string
  quote_symbol: string | null
  supported: number
  description: string | null
  created_at: string
  updated_at: string
}

export interface IndexQuoteSymbolPage {
  items: IndexQuoteSymbol[]
  total: number
  limit: number
  offset: number
}

export async function refreshMarketQuotes(): Promise<RefreshMarketResult> {
  const { data } = await apiClient.post<RefreshMarketResult>('/market/refresh', undefined, {
    timeout: 120000,
  })
  return data
}

export async function listIndexQuoteSources(): Promise<IndexQuoteSourceStatus[]> {
  const { data } = await apiClient.get<IndexQuoteSourceStatus[]>('/market/index-quote-sources')
  return data
}

export interface IndexQuoteSourceRuleUpdate {
  source_description?: string | null
  exclude_rule_type: string
  exclude_rule_value?: string | null
}

export async function updateIndexQuoteSource(
  sourceKey: string,
  payload: IndexQuoteSourceRuleUpdate,
): Promise<IndexQuoteSourceStatus> {
  const { data } = await apiClient.put<IndexQuoteSourceStatus>(`/market/index-quote-sources/${sourceKey}`, payload)
  return data
}

export interface IndexQuoteSymbolUpdate {
  index_code: string
  source_key: string
  quote_symbol?: string | null
  supported: number
  description?: string | null
}

export interface IndexQuoteSymbolQuery {
  index_code?: string
  source_key?: string
  limit?: number
  offset?: number
}

export async function listIndexQuoteSymbols(params: IndexQuoteSymbolQuery = {}): Promise<IndexQuoteSymbolPage> {
  const { data } = await apiClient.get<IndexQuoteSymbolPage>('/market/index-quote-symbols', { params })
  return data
}

export async function upsertIndexQuoteSymbol(payload: IndexQuoteSymbolUpdate): Promise<IndexQuoteSymbol> {
  const { data } = await apiClient.put<IndexQuoteSymbol>('/market/index-quote-symbols', payload)
  return data
}
